File: src/lib/threat.ts
import { now } from './clock.js';
import { distanceTo, hex, nameOf } from './entity.js';
import { hitsCeiling } from './vitals.js';

export interface ThreatOptions {
  prefix: string;
  range: number;

  // A SearchEntityOptions bitmask
  hostile: number;

  companion?: () => Mobile | undefined;
  companionName: string;

  call: string;
  calls: number;
  callDelay: number;
  replyWait: number;
  noGuardsText: string[];
  attackText: string[];
  guardedText: string[];
  unguardedText: string[];
}

export interface ThreatWatch {
  check: () => void;
}

// Notorieties, so the line that opens an episode says why the thing matched HOSTILE_NOTORIETY
const NOTORIETY = [
  'unknown',
  'innocent',
  'ally',
  'gray',
  'criminal',
  'enemy',
  'murderer',
  'invulnerable',
];

// 0 is what the client reports while it is refreshing stats, and what it reports for a mobile it has
// lost track of, so a fall to 0 is no news at all.
const dropped = (was: number, is: number): boolean => was > 0 && is > 0 && is < was;

export const createThreatWatch = (options: ThreatOptions): ThreatWatch => {
  let lastHits = 0;
  let lastCompanionHits = 0;
  let lastCall = 0;
  let calls = 0;
  let episode = false;
  let noGuards = false;
  let saidProtection = false;
  let zone: string | undefined;

  const nearest = (mask: number, type: number): Mobile | undefined => {
    const found = client.selectEntity(mask, SearchEntityRangeOptions.Nearest, type, false);

    if (!found || found.serial === player.serial || found.isDead) {
      return undefined;
    }

    return distanceTo(found) <= options.range ? found : undefined;
  };

  // isRenamable is how the rest of the repo tells your own pet from a stranger's, and a pet flagged
  // gray by whatever it was fighting would otherwise read as the thing attacking you.
  const hostileNear = (): Mobile | undefined => {
    const found = nearest(options.hostile, SearchEntityTypeOptions.Any);

    return found && !found.isRenamable ? found : undefined;
  };

  const readZone = (): void => {
    if (options.guardedText.some((text) => journal.containsText(text))) {
      zone = 'guarded';
    } else if (options.unguardedText.some((text) => journal.containsText(text))) {
      zone = 'unguarded';
    }
  };

  // Nothing in the API answers this. A yellow human is a guard or a vendor, and either one means a
  // town, which is the best the client can be asked.
  const protection = (): string => {
    if (zone) {
      return `the journal says ${zone}`;
    }

    const yellow = nearest(SearchEntityOptions.Invulnerable, SearchEntityTypeOptions.Human);

    return yellow
      ? `an invulnerable '${nameOf(yellow)}' in sight, so probably a town`
      : 'nothing in sight to say either way';
  };

  const callGuards = (): void => {
    if (noGuards || (options.calls > 0 && calls >= options.calls)) {
      return;
    }

    const at = now();

    if (calls > 0 && at - lastCall < options.callDelay) {
      return;
    }

    lastCall = at;
    calls++;

    if (!saidProtection) {
      saidProtection = true;
      log(`${options.prefix}: guard protection - ${protection()}`);
    }

    log(
      `${options.prefix}: calling the guards (${calls}${options.calls > 0 ? `/${options.calls}` : ''})`,
    );
    player.say(options.call);

    if (options.noGuardsText.length === 0) {
      return;
    }

    const refused = journal.waitForTextAny(options.noGuardsText, undefined, options.replyWait);

    if (refused) {
      noGuards = true;
      log(`${options.prefix}: the shard says '${refused}' - not calling again this run`);
    }
  };

  const describe = (hostile: Mobile | undefined, friend: Mobile | undefined): string => {
    const who = hostile
      ? `'${nameOf(hostile)}' ${hex(hostile.graphic)} ${distanceTo(hostile)} tiles off ` +
        `(${NOTORIETY[hostile.notoriety] ?? hostile.notoriety})`
      : 'nothing in sight';

    const mine = `you ${player.hits}/${hitsCeiling() ?? '?'}`;
    const theirs = friend
      ? `, ${options.companionName} ${friend.hits}/${friend.maxHits || '?'}`
      : '';

    return `${who}, ${mine}${theirs}`;
  };

  return {
    check: () => {
      readZone();

      const hits = player.hits;
      const hurt = dropped(lastHits, hits);

      if (hits > 0) {
        lastHits = hits;
      }

      const friend = options.companion?.();
      const friendHits = friend?.hits ?? 0;
      const friendHurt = dropped(lastCompanionHits, friendHits);

      if (friendHits > 0) {
        lastCompanionHits = friendHits;
      }

      const said = options.attackText.some((text) => journal.containsText(text));
      const hostile = hostileNear();

      if (!hostile && !hurt && !friendHurt && !said) {
        if (episode) {
          episode = false;
          calls = 0;
          log(`${options.prefix}: clear`);
        }

        return;
      }

      if (!episode) {
        episode = true;
        log(`${options.prefix}: trouble - ${describe(hostile, friend)}`);
      }

      callGuards();
    },
  };
};

File: legion/src/uo/threat.py
import API

from uo.alert import Launcher
from uo.entity import hex_of
from uo.journal import forget, matched_bucket
from uo.notoriety import HOSTILE


# 0 is what the client reports while it is refreshing stats, and for a mobile it has lost track of,
# so a fall to 0 is no news at all
def dropped(was, is_now):
    return was > 0 and is_now > 0 and is_now < was


def hostiles_near(notoriety, within):
    found = API.GetAllMobiles(None, within, notoriety) or []

    for mobile in found:
        # IsRenamable is how the rest of this repo tells your own pet from a stranger's, and a pet
        # flagged gray by whatever it was fighting would otherwise read as the thing attacking you
        if mobile.Serial != API.Player.Serial and not mobile.IsDead and not mobile.IsRenamable:
            return mobile

    return None


class ThreatWatch(object):
    def __init__(self, config, log, companion, friend_label, hold=None):
        self._config = config
        self._log = log
        self._companion = companion
        self._friend_label = friend_label
        self._hold = hold
        self._alert = Launcher(log)
        self._last_hits = 0
        self._last_companion_hits = 0
        self._in_episode = False
        self._trouble_seen = False
        self._alarm_left = 0

    # Consuming: the roam idle loop never clears the journal, so said() would re-arm this every poll
    def _ambushed(self):
        text = self._config["ambush_text"]
        return bool(text) and matched_bucket([("ambushed", text)]) is not None

    def _sound(self):
        if self._alarm_left > 0 and self._alert.play(self._config["ambush_alarm"]):
            self._alarm_left -= 1

    def _describe(self, hostile, friend):
        if hostile is not None:
            who = "'%s' %s %d tiles off" % (
                hostile.Name or "?",
                hex_of(hostile.Graphic),
                hostile.Distance,
            )
        else:
            who = "nothing in sight"

        ceiling = API.Player.HitsMax
        mine = "you %d/%s" % (API.Player.Hits, ceiling if ceiling > 0 else "?")
        theirs = ""

        if friend is not None:
            theirs = ", %s %d/%s" % (self._friend_label(friend), friend.Hits,
                                     friend.HitsMax or "?")

        return "%s, %s%s" % (who, mine, theirs)

    def look(self):
        if not self._config["watch"]:
            return

        hits = API.Player.Hits
        hurt = dropped(self._last_hits, hits)

        if hits > 0:
            self._last_hits = hits

        friend = self._companion()
        friend_hits = friend.Hits if friend is not None else 0
        friend_hurt = dropped(self._last_companion_hits, friend_hits)

        if friend_hits > 0:
            self._last_companion_hits = friend_hits

        hostile = hostiles_near(HOSTILE, self._config["range"])
        trouble = hostile is not None or hurt or friend_hurt

        if self._ambushed():
            self._in_episode = True
            self._trouble_seen = False
            self._alarm_left = self._config["ambush_repeats"] if self._config["ambush_alarm"] else 0
            self._log("ambushed - %s" % self._describe(hostile, friend))
            API.HeadMsg(self._config["ambush_warning"], API.Player.Serial,
                        self._config["ambush_hue"])

            for command in self._config["ambush_notices"]:
                self._alert.run(command)

            # The scan above is stale once the hold returns; the next look reads the fight afresh
            if self._hold is not None:
                self._hold.wait(self._sound)
                # The button means the player has judged it safe, so every ambush line from while
                # the gump was up is stale - including the hold's own log line, which the client
                # files in the journal and which carries the very phrase this watch waits on
                forget(self._config["ambush_text"])
                self._in_episode = False
                self._trouble_seen = False
                self._alarm_left = 0
                self._alert.stop()

                return

        if trouble:
            if not self._in_episode:
                self._in_episode = True
                self._log("trouble - %s" % self._describe(hostile, friend))

            self._trouble_seen = True
        # An ambush announces monsters that take a cycle to appear, so the alarm outlives an empty
        # scan until a fight has come and gone or the repeats run out
        elif self._in_episode and (self._trouble_seen or self._alarm_left == 0):
            self._in_episode = False
            self._trouble_seen = False
            self._alarm_left = 0
            self._alert.stop()
            self._log("clear")

        self._sound()
